### PR TITLE
[Link Handling] Make link references consistent and flexible #713 - v2

### PR DESCRIPTION
**Pull request by stefanseifert**
- introduce Link interface for consistent representation of link target information in all model APIs
- introduce a simple (and internal) LinkHandler implementation to resolve all types of links in all implementations, deprecate Utils methods for the same purpose
- prepare support for link target in the model implementations
- create new version of all models that are affected by the new link handling
  - breadcrumb v2
  - button v2
  - image v3
  - languagenavigation v2
  - list v3
  - navigation v2
  - teaser v2
  - title v3
  - page v3
- the new model versions contain on the new "link" objects in JSON output instead of the old inconsistently named attributes containing the link URL

the PR also contains two small non-functional improvements in separate commits:
- compare JSON results in unit tests with jsonassert to get better reporting about differences if the test fails
- ignore findbugs plugin execution in eclipse

fixes #713

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/commons/link/Link.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     * @return Link URL or null if link is invalid.
+     */
+    @Nullable
+    String getURL();
+
+    /**
+     * Map with Attributes for HTML Anchor tag for this links.
+     * This usually also contains the Link URL as <code>href</code> attribute,
+     * but may contain additional attributes like <code>target</code> and others.
+     * @return Map with HTML-specific anchor attributes, or null if link is invalid
```
Comment (kwin) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/commons/link/Link.java]: may return null but annotated with `@NotNull`

**Comment by stefanseifert** [comment]
the complaints about the broken builds result that this branch is outdated compared to the development/master branch, and there was a new release in between which added new methods to the public APIs missing in this branch.

once the branch is synchronized with the current development branch these failures go away (but please not this is much more effort than doing a git merge because all changes to the AEM component definitions, HTL files etc. have to be re-applied manually to the copies in the new component versions as warned in this mailing list thread
https://groups.google.com/forum/#!msg/aem-core-components-dev/ITKYKRjqRBI/0x76fZMBCQAJ

**Comment by kwin** [comment]
IIUC then https://github.com/adobe/aem-core-wcm-components/pull/862/files#diff-13ccf10ea7dbf29d5ba83f8ee9bad001R43 won't return the mapped URL. That is necessary though for some use cases which are not processed by the LinkRewriter (e.g. https://github.com/adobe/aem-core-wcm-components/blob/master/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java#L120). I would propose to add a 2nd method which returns the mapped URL or alternatively a boolean flag as argument for `getUrl()` which determines whether the returned link should be mapped or not!

**Comment by stefanseifert** [comment]
this PR is only the first step to introduced centralized link handling logic in core components at all. once this is done there are lots of more features that should make the link handling more robust and flexible (like your previous comment, and like #725 and #727).

i'm waiting until this PR get's merged (which is spanning over 6 months now including the previous one) until investing more efforts...

**Comment by stefanseifert** [comment]
@bpauli thanks for merging this PR!

are you aware that simply merging it is not enough, all files for the new versions of each affected component needs to be updated as well, all changes to the HTL script, dialog definitions, component definitions which where maid since november on the current versions need to be manually re-applied to the 3.0.0 branch (and this has to be done in the future as well until 3.0.0 is released and the main work focuses on the new version of the components).

**Comment by ky940819** [comment]
Since a lot of new component versions are being added for this issue - would it be possible to implement #1254 with these new versions.

The reason given for not changing the behavior was the possibility of clients having component models that do not implement the Component interface; however, with new versions of the components there should be no concern about this.

This would present a good opportunity for optimization, by eliminating the need to adapt the request -> Component twice.

**Comment by vladbailescu** [comment]
@ky940819 , yes, we'll try to make sure each newly introduced component version is using the `id` from it's own model directly.

**Pull request merged**